logger/src/lib.rs

```diff
@@ -1,159 +1,139 @@
-use tracing_bunyan_formatter::{BunyanFormattingLayer, JsonStorageLayer};
 pub use tracing_futures::Instrument as InstrumentFutures;
+pub use tracing_subscriber::reload::Error as ReloadError;
 use tracing_subscriber::{layer::SubscriberExt, registry::Registry, reload};
 
-/// Substrate telemetry
-pub type SubstrateTelemetry = Receiver<Telemetry>;
-
-/// Future telemetry
-pub type FutureTelemetry = Receiver<Telemetry>;
-
-/// Convenience wrapper for Telemetry types.
-pub type Telemetries = (SubstrateTelemetry, FutureTelemetry);
+const TELEMETRY_CAPACITY: usize = 1000;
 
 static LOGGER_SET: AtomicBool = AtomicBool::new(false);
 
-/// Initializes `Logger` with given [`Configuration`].
-/// After the initialization `log` macros will print with the use of this `Logger`.
-/// Returns the receiving side of telemetry channels (regular telemetry, future telemetry)
-///
-/// # Errors
-/// If the logger is already set, raises a generic error.
-pub fn init(configuration: &Configuration) -> Result<Option<Telemetries>> {
+fn try_set_logger() -> Result<()> {
     if LOGGER_SET
         .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
         .is_err()
     {
-        return Ok(None);
+        return Err(eyre!("Logger is already set."));
     }
-    Ok(Some(setup_logger(configuration)?))
+    Ok(())
 }
 
-/// Disables the logger by setting `LOGGER_SET` to true. Will fail
-/// if the logger has already been initialized. This function is
-/// required in order to generate flamegraphs and flamecharts.
+/// Initializes the logger globally with given [`Configuration`].
 ///
-/// Returns true on success.
-pub fn disable_logger() -> bool {
-    LOGGER_SET
-        .compare_exchange(false, true, Ordering::SeqCst, Ordering::SeqCst)
-        .is_ok()
-}
+/// Returns [`LoggerHandle`] to interact with the logger instance
+///
+/// Works only once per process, all subsequent invocations will fail.
+///
+/// For usage in tests consider [`test_logger`].
+///
+/// # Errors
+/// If the logger is already set, raises a generic error.
+// TODO: refactor configuration in a way that `terminal_colors` is part of it
+//       https://github.com/hyperledger/iroha/issues/3500
+pub fn init_global(configuration: &Configuration, terminal_colors: bool) -> Result<LoggerHandle> {
+    try_set_logger()?;
 
-fn setup_logger(configuration: &Configuration) -> Result<Telemetries> {
     let layer = tracing_subscriber::fmt::layer()
-        .with_ansi(configuration.terminal_colors)
+        .with_ansi(terminal_colors)
         .with_test_writer();
 
-    if configuration.compact_mode {
-        add_bunyan(configuration, layer.compact())
-    } else {
-        add_bunyan(configuration, layer)
+    match configuration.format {
+        Format::Full => step2(configuration, layer),
+        Format::Compact => step2(configuration, layer.compact()),
+        Format::Pretty => step2(configuration, layer.pretty()),
+        Format::Json => step2(configuration, layer.json()),
```

Comment:
I _guess_ `bunyan` might read `tracing`'s JSON logs without issues (I will check). `tracing` docs [says](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/format/struct.Json.html) it might be used by some `jq` or a JSON log viewer.

Anyway, we should communicate this change to users, for sure. They will need to change logger configuration anyway.

Reply:
Nope, `bunyan` cannot read it: [it requires at least the `name` field to be set](https://github.com/trentm/node-bunyan#introduction), which `tracing` doesn't provide.